### PR TITLE
t/ckeditor5/1457: Implemented Font Color and Font Background Color features

### PR DESCRIPTION
**Pull request by msamsel**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Implemented Font Color and Font Background Color features. Closes ckeditor/ckeditor5#1457.

---

### Additional information

* CI: https://github.com/ckeditor/ckeditor5/compare/t/1457?expand=1

PRs:
* https://github.com/ckeditor/ckeditor5-core/pull/169
* https://github.com/ckeditor/ckeditor5-highlight/pull/38
* https://github.com/ckeditor/ckeditor5-theme-lark/pull/222
* https://github.com/ckeditor/ckeditor5-ui/pull/474

### CODE REVIEW

**Review by mlewand — COMMENTED** [review]
Couple of early review notes:

- [x] Remove color button does not use "pointer" cursor upon hover
- [x] I don't think you're supposed to be able to click into inactive icon (it shows the outline, despite not doing anything). I believe it's an upstream, is it already tracked somewhere @msamsel?
- [x] The default color palette must come with (localized) labels.
- [ ] Manual test should include some description.

**Comment by msamsel** [comment]
> I don't think you're supposed to be able to click into inactive icon (it shows the outline, despite not doing anything). I believe it's an upstream, is it already tracked somewhere @msamsel?

I've report it here as I couldn't find similar issue: https://github.com/ckeditor/ckeditor5-ui/issues/473.

File: src/fontbackgroundcolor/fontbackgroundcolorcommand.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module font/fontbackgroundcolor/fontbackgroundcolorcommand
+ */
+
+import FontCommand from '../fontcommand';
+import { FONT_BACKGROUND_COLOR } from '../utils';
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorcommand.js:11]: Code style: missing new line.

File: src/fontbackgroundcolor/fontbackgroundcolorediting.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module font/fontfamily/fontfamilyediting
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorediting.js]: Code style: too many new lines :p

File: src/fontbackgroundcolor/fontbackgroundcolorediting.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module font/fontfamily/fontfamilyediting
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorediting.js]: Docs: Wrong path.

File: src/fontbackgroundcolor/fontbackgroundcolorui.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * @module font/fontbackgroundcolor/fontbackgroundolorui
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+
+import fontBackgroundColorIcon from '../../theme/icons/font-background.svg';
+import { createDropdown } from '@ckeditor/ckeditor5-ui/src/dropdown/utils';
+import { FONT_BACKGROUND_COLOR, normalizeOptions, colorUI } from '../utils';
+export default class FontBackgroundColorUI extends Plugin {
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorui.js]: Code style: missing new line.

File: src/fontbackgroundcolor/fontbackgroundcolorui.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+		const editor = this.editor;
+		const t = editor.t;
+		const command = editor.commands.get( FONT_BACKGROUND_COLOR );
+
+		const options = this._getLocalizedOptions();
+
+		// Register UI component.
+		editor.ui.componentFactory.add( FONT_BACKGROUND_COLOR, locale => {
+			const dropdownView = createDropdown( locale );
+			const colorTableView = colorUI.addColorsToDropdown(
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorui.js]: Why do we need this `ColorUI namespace`? It holds just a single function? Why not a plain helper instead?

File: src/fontbackgroundcolor/fontbackgroundcolorui.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+				dropdownView,
+				options.map( element => ( {
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorui.js]: Destructuring?

File: src/fontbackgroundcolor/fontbackgroundcolorui.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+			dropdownView.extendTemplate( {
+				attributes: {
+					class: 'ck-font-background-color-dropdown'
+				}
+			} );
+
+			dropdownView.bind( 'isEnabled' ).to( command );
+
+			dropdownView.on( 'execute', ( evt, val ) => {
+				if ( val.value !== null ) {
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorui.js]: `val.value` looks weird. There's something wrong with the naming convention somewhere if it comes to that.

File: src/fontbackgroundcolor/fontbackgroundcolorui.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+					colorTableView.recentlyUsedColors.add( { color: val.value, hasBorder: val.hasBorder, label: val.label }, 0 );
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorui.js]: Let's put the props in separate lines. It's hard to read.

File: src/fontbackgroundcolor/fontbackgroundcolorui.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+	 * Note: The reason behind this method is that there is no way to use {@link module:utils/locale~Locale#t}
+	 * when the user configuration is defined because the editor does not exist yet.
+	 *
+	 * @private
+	 * @returns {Array.<module:font/fontfamily~FontFamilyOption>}.
+	 */
+	_getLocalizedOptions() {
+		const editor = this.editor;
+		const colors = normalizeOptions( editor.config.get( `${ FONT_BACKGROUND_COLOR }.colors` ) );
+		return colors;
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorui.js]: `return normalizeOptions(...`

File: src/fontbackgroundcolor/fontbackgroundcolorui.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+				editor.execute( FONT_BACKGROUND_COLOR, val );
+				editor.editing.view.focus();
+			} );
+
+			return dropdownView;
+		} );
+	}
+
+	/**
+	 * Returns options as defined in `config.fontFamily.options` but processed to account for
```
Comment (oleq) [inline src/fontbackgroundcolor/fontbackgroundcolorui.js]: This entire doc string should be about font bg color, not font family.

File: src/fontcolor/fontcolorcommand.js [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module font/fontcolor/fontcolorcommand
+ */
+
+import FontCommand from '../fontcommand';
+import { FONT_COLOR } from '../utils';
+export default class FontColorCommand extends FontCommand {
```
Comment (oleq) [inline src/fontcolor/fontcolorcommand.js:23]: Missing new line.

File: src/fontcolor/fontcolorediting.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module font/fontfamily/fontfamilyediting
```
Comment (oleq) [inline src/fontcolor/fontcolorediting.js]: Wrong path.

File: src/fontcolor/fontcolorediting.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+
```
Comment (oleq) [inline src/fontcolor/fontcolorediting.js]: Too many new lines.

File: src/fontcolor/fontcolorediting.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+
+		// Allow fontColor attribute on text nodes.
+		editor.model.schema.extend( '$text', { allowAttributes: FONT_COLOR } );
```
Comment (oleq) [inline src/fontcolor/fontcolorediting.js:119]: Out of curiosity: why it is that this line must be in `init()` and not in the constructor? Does it have something to do with other editor plugins?

File: src/fontcolor/fontcolorui.js [hunk 15]
```diff
@@ -0,0 +1,85 @@
+/**
```
Comment (oleq) [inline src/fontcolor/fontcolorui.js:1]: TBH this entire files is just a copy of `fontbackgroundcolorui.js`, right? Maybe let's have them both use the same UI component then? DRY.

File: src/ui/colorgrid.js [hunk 16]
```diff
@@ -0,0 +1,93 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
```
Comment (oleq) [inline src/ui/colorgrid.js]: No `@module` docs.

File: src/ui/colorgrid.js [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+		colorsDefinition.forEach( item => {
+			const colorTile = new ColorTile();
+			colorTile.set( {
+				color: item.color,
+				label: item.label,
+				tooltip: true,
+				hasBorder: item.options.hasBorder
+			} );
+			colorTile.on( 'execute', () => {
+				this.fire( 'execute', { value: item.color, hasBorder: item.options.hasBorder, label: item.label } );
```
Comment (oleq) [inline src/ui/colorgrid.js]: Please avoid object long single–line obj literals.

File: src/ui/colortableview.js [hunk 18]
```diff
@@ -0,0 +1,164 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
```
Comment (oleq) [inline src/ui/colortableview.js:5]: No `@module` docs.

File: src/ui/colortableview.js [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+import View from '@ckeditor/ckeditor5-ui/src/view';
+import ButtonView from '@ckeditor/ckeditor5-ui/src/button/buttonview';
+import ColorTile from './colortile';
+import ColorGrid from './colorgrid';
+
+import removeButtonIcon from '@ckeditor/ckeditor5-core/theme/icons/eraser.svg';
```
Comment (oleq) [inline src/ui/colortableview.js]: Generally we import js first, then svg and css.

File: src/ui/colortableview.js [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+		this.locale = locale;
+		this.items = this.createCollection();
+
+		this.colorsDefinition = colors;
+		this.focusTracker = new FocusTracker();
+		this.keystrokes = new KeystrokeHandler();
+
+		this.set( 'selectedColor' );
+		this.set( 'hoveredColor' );
+		this.set( 'removeButtonTooltip' );
```
Comment (oleq) [inline src/ui/colortableview.js]: Why is this prop observable? Does it change at some point?

File: src/ui/colortableview.js [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+		super( locale );
+		this.locale = locale;
+		this.items = this.createCollection();
+
+		this.colorsDefinition = colors;
+		this.focusTracker = new FocusTracker();
+		this.keystrokes = new KeystrokeHandler();
+
+		this.set( 'selectedColor' );
+		this.set( 'hoveredColor' );
```
Comment (oleq) [inline src/ui/colortableview.js]: I don't see this prop used anywhere.

File: src/ui/colortableview.js [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+		this.set( 'removeButtonTooltip' );
+		this.set( 'colorColumns', 5 );
```
Comment (oleq) [inline src/ui/colortableview.js]: Why is this prop observable? It's a constant isn't it?

File: src/ui/colortableview.js [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+		this.set( 'recentlyUsedColors', new Collection() );
```
Comment (oleq) [inline src/ui/colortableview.js]: Why is this prop observable? Is the collection re–initialised from scratch at some point?

File: src/ui/colortableview.js [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+					hasBorder: colorObj.hasBorder
+				} );
+				if ( colorObj.label ) {
+					colorTile.set( {
+						label: colorObj.label,
+						tooltip: true
+					} );
+				}
+				if ( colorObj.isEnabled === false ) {
+					colorTile.set( 'isEnabled', false );
```
Comment (oleq) [inline src/ui/colortableview.js:207]: `colorTile.set( 'isEnabled', colorObj.isEnabled );` ?

File: src/ui/colortile.js [hunk 25]
```diff
@@ -0,0 +1,28 @@
+/**
+ * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import ButtonView from '@ckeditor/ckeditor5-ui/src/button/buttonview';
```
Comment (oleq) [inline src/ui/colortile.js]: Missing new line.

File: src/utils.js [hunk 26]
```diff
@@ -35,3 +37,70 @@ export function buildDefinition( modelAttributeKey, options ) {
...
+	return value.replace( /\s/g, '' );
+}
+
+export function normalizeOptions( colorRow ) {
+	return colorRow
+		.map( getColorsDefinition )
+		.filter( option => !!option );
+}
+
+function getColorsDefinition( color ) {
```
Comment (oleq) [inline src/utils.js]: `getColorDefinition` if anything. But this has something to do with the conversion — maybe there's a better name for it?

File: src/utils.js [hunk 27]
```diff
@@ -35,3 +37,70 @@ export function buildDefinition( modelAttributeKey, options ) {
...
+				styles: {
+					color: `${ color.color }`
+				},
+				priority: 5
+			}
+		};
+	}
+}
+
+export const colorUI = {
```
Comment (oleq) [inline src/utils.js]: As I wrote earlier, this namespace is pointless with a single function like this.

**Review by oleq — CHANGES_REQUESTED** [review]
The feature looks promising 🎉 but there are some things that needs fixing.

P.S. I didn't point it out inline but almost every prop, method, class and config are missing docs.

We need a feature guide in https://ckeditor.com/docs/ckeditor5/latest/features/index.html too.

File: theme/fontcolor.css [hunk 28]
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+:root {
+	--ck-table-color-tile-size: 20px;
+	--ck-table-color-max-columns: 5;
+	--ck-table-color-padding: 10px;
```
Comment (oleq) [inline theme/fontcolor.css]: Please use constants from `_spacing.css` instead.

File: theme/fontcolor.css [hunk 29]
```diff
@@ -0,0 +1,61 @@
...
+	padding: var(--ck-table-color-padding);
+}
+
+.ck .ck-color-table__color-tile {
+	width: var(--ck-table-color-tile-size);
+	height: var(--ck-table-color-tile-size);
+	min-width: var(--ck-table-color-tile-size);
+	min-height: var(--ck-table-color-tile-size);
+	border-radius: var(--ck-border-radius);
+	cursor: pointer;
```
Comment (oleq) [inline theme/fontcolor.css]: This is a wrong cursor. A default one should be used instead. Pointer is reserved for links.

File: theme/fontcolor.css [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+	grid-gap: calc( var(--ck-table-color-padding) / 2 );
+	padding: var(--ck-table-color-padding);
+}
+
+.ck .ck-color-table__color-tile {
+	width: var(--ck-table-color-tile-size);
+	height: var(--ck-table-color-tile-size);
+	min-width: var(--ck-table-color-tile-size);
+	min-height: var(--ck-table-color-tile-size);
+	border-radius: var(--ck-border-radius);
```
Comment (oleq) [inline theme/fontcolor.css]: Use `@mixin ck-rounded-corners` instead.

File: theme/fontcolor.css [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+/* If we change it to list element this CSS will be obsolete. */
+.ck .ck-button.ck-color-table__remove-color {
+	display: flex;
+	align-items: center;
+	width: 100%;
+	border-bottom: 1px solid var(--ck-color-base-border);
+	padding: calc(var(--ck-table-color-padding) / 2 ) var(--ck-table-color-padding);
+}
+
+.ck .ck-button.ck-button_with-text.ck-color-table__remove-color svg {
```
Comment (oleq) [inline theme/fontcolor.css]: What is wrong with default `.ck-button` styles? They are shared across the project.

File: theme/fontcolor.css [hunk 32]
```diff
@@ -0,0 +1,61 @@
...
+	width: var(--ck-table-color-tile-size);
+	margin-right: var(--ck-spacing-large);
+}
+
+.ck .ck-color-table__remove-color {
+	cursor: pointer;
```
Comment (oleq) [inline theme/fontcolor.css]: This is wrong and does not even work BTW :P

**Review by oleq — CHANGES_REQUESTED** [review]
Some changes in styles are required too.

**Comment by oleq** [comment]
Also

```
(master 863a704) oleq@MBP15 ~/ck/5/ckeditor5> yarn docs --skip-snippets
yarn run v1.13.0
$ node ./scripts/docs/build-docs.js --skip-snippets
JSDoc started...
Memberof property should start with 'module:'. Got 'ColorTableView' instead.
	at /Users/oleq/ck/5/ckeditor5/packages/ckeditor5-font/src/ui/colortableview.js (line 137)


Validation Summary:
 * Validator found 1 errors.
Documented 421 files!
Clean build directory.
Emptying Hexo source directory.
Building API docs of CKEditor 5...
Built API docs in: 9614.659ms
Error: Invalid internal links in /ckeditor5/12.0.0/api/module_font_fontbackgroundcolor_fontbackgroundcolorcommand.html:
  * 'module_font_fontbackgroundcolor_fontbackgroundcolorcommand.html#FontBackgroundColorCommand' - FontBackgroundColorCommand
Error: Invalid internal links in /ckeditor5/12.0.0/api/module_font_fontcolor_fontcolorcommand.html:
  * 'module_font_fontcolor_fontcolorcommand.html#FontColorCommand' - FontColorCommand
Error: Invalid internal links in /ckeditor5/12.0.0/api/module_font_fontfamily_fontfamilyediting.html:
  * 'module_font_fontfamily_fontfamilyediting.html#FontBackgroundColorEditing' - FontBackgroundColorEditing
  * 'module_font_fontfamily_fontfamilyediting.html#FontColorEditing' - FontColorEditing
Running webpack for app.js ...
```

File: theme/fontcolor.css (hunk 31 above)
Comment (oleq) [reply theme/fontcolor.css]: If it looks like they need fixing, this should be done somewhere in `button.css` instead, so the entire project benefits from the fix. Do we use this kind of button anywhere anyway?

Comment (oleq) [reply theme/fontcolor.css]: Check out [theme lark MT](http://localhost:8125/ckeditor5-theme-lark/tests/manual/theme.html). It should look good without any additional tuning

image

cc @dkonopka

File: src/fontcolor/fontcolorediting.js (hunk 14 above)
Comment (msamsel) [reply src/fontcolor/fontcolorediting.js:119]: Not really. I just take inspiration from other plugin.

File: src/fontbackgroundcolor/fontbackgroundcolorui.js (hunk 6 above)
Comment (msamsel) [reply src/fontbackgroundcolor/fontbackgroundcolorui.js]: I'm not sure what you have on your mind here. There is returned array with objects, where I only take few of object's properties. Can you give me more hints what you had on your mind?

File: src/ui/colortableview.js (hunk 24 above)
Comment (msamsel) [reply src/ui/colortableview.js:207]: It have to be resist for case when `colorObj.isEnabled` is undefined. However if you think that that better is to compress it, I could make sure that `colorObj.isEnabled` will be alway set up to true or false.

**Comment by jodator** [comment]
My 2 cents:
- consider different keyboard navigation #26 
- white color has no focus outline on keyboard navigation:

    Peek 2019-03-28 18-53
    **(the colors on buttons were only for testing purposes)**
- currently selected color has no focus outline on keyboard navigation:

    Peek 2019-03-28 18-54

- I don't like `FontBackgroundColor*` name. For me `FontBackground*` would be enough :) But it's rather meh..

**Comment by jodator** [comment]
OK. I think I'm done with reading all that :D

**Comment by mlewand** [comment]
## Corrections

> The default value registers 15 colors:

changed to:

> The default value registers following colors:

It's better not to mention too much details of value that is right below, that way we reduce thw chance that someone edits the listing alone (adds/removes some colors) without updating this particular line.

I have simplified manual test description (it was basically duplicating the editable contnet).

## Issues

* The "recent/document" colors feature to pick up pasted/preloaded colors is missing, but this should not be blocking this PR extracted to #28.
* `hasBorder` property won't work properly, as here the assumption is that the skin will be bright (as it's applied to bright colors). It won't work without reconfiguration with dark themes.

minor:

* There's no string value example in `FontBackgroundColorConfig.colors` (and respective font color struct).
* FontBackgroundColorConfig and FontColorConfig implement identical interface and both have the same default values. It would improve DRY factor to have a common interface for both features. I see that now the listings differ in both, but having common interface should simply mention both configuration names.


Above things will be extracted to a separate issues not to block this PR.

**Review by mlewand — APPROVED** [review]
Ok all the things extracted, PR is ready to land on master branch. I'll merge it in few mins.

**Pull request merged**